File: packages/sdk-php/tests/ZenvioTest.php
<?php

namespace Zenvio\Tests;

use PHPUnit\Framework\TestCase;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use GuzzleHttp\Middleware;
use Zenvio\Zenvio;
use Zenvio\Exception\ZenvioApiException;

class ZenvioTest extends TestCase
{
    private array $container = [];

    private function getMockClient(array $responses): Zenvio
    {
        $mock = new MockHandler($responses);
        $history = Middleware::history($this->container);
        $handlerStack = HandlerStack::create($mock);
        $handlerStack->push($history);
        return new Zenvio('test-key', 'https://api.zenvio.com/v1', ['handler' => $handlerStack]);
    }

    public function testWhatsAppSendText(): void
    {
        $zenvio = $this->getMockClient([
            new Response(200, [], json_encode(['message_ids' => ['msg-123'], 'status' => 'queued'])),
        ]);
        $resp = $zenvio->whatsapp->sendText('instance-1', ['5511999999999'], 'Hello');
        $this->assertIsArray($resp);
        $this->assertSame(['msg-123'], $resp['message_ids']);
        $this->assertSame('queued', $resp['status']);
    }

    public function testWhatsAppSendErrorThrows(): void
    {
        $zenvio = $this->getMockClient([
            new Response(400, [], json_encode(['error' => 'Invalid instance'])),
        ]);
        $this->expectException(ZenvioApiException::class);
        $zenvio->whatsapp->sendText('x', ['123'], 'hi');
    }

    public function testWhatsAppSendWithParams(): void
    {
        $zenvio = $this->getMockClient([
            new Response(200, [], json_encode(['message_ids' => ['m1'], 'status' => 'queued'])),
        ]);
        $resp = $zenvio->whatsapp->send('instance-1', [
            'to' => ['5511888888888'],
            'type' => 'text',
            'payload' => ['message' => 'Hi'],
        ]);
        $this->assertSame(['m1'], $resp['message_ids']);
    }

    public function testWhatsAppGetMessage(): void
    {
        $zenvio = $this->getMockClient([
            new Response(200, [], json_encode(['message_id' => 'msg-1', 'status' => 'delivered'])),
        ]);
        $resp = $zenvio->whatsapp->getMessage('msg-1');
        $this->assertSame('msg-1', $resp['message_id']);
        $this->assertSame('delivered', $resp['status']);
    }

    public function testWhatsAppListInstances(): void
    {
        $zenvio = $this->getMockClient([
            new Response(200, [], json_encode(['success' => true, 'data' => [], 'pagination' => ['total' => 0]])),
        ]);
        $resp = $zenvio->whatsapp->listInstances();
        $this->assertTrue($resp['success']);
    }

    public function testSmsSend(): void
    {
        $zenvio = $this->getMockClient([
            new Response(200, [], json_encode([
                'success' => true,
                'data' => ['status' => 'queued', 'count' => 1, 'sms_ids' => ['sms-1']],
            ])),
        ]);
        $resp = $zenvio->sms->send(['to' => ['5511999999999'], 'message' => 'Test']);
        $this->assertTrue($resp['success']);
        $this->assertSame(['sms-1'], $resp['data']['sms_ids']);
    }

    public function testSmsGet(): void
    {
        $zenvio = $this->getMockClient([
            new Response(200, [], json_encode([
                'success' => true,
                'data' => ['sms_id' => 'sms-1', 'status' => 'delivered'],
            ])),
        ]);
        $resp = $zenvio->sms->get('sms-1');
        $this->assertSame('sms-1', $resp['data']['sms_id']);
    }

    public function testSmsCancel(): void
    {
        $zenvio = $this->getMockClient([
            new Response(200, [], json_encode([
                'success' => true,
                'data' => ['sms_id' => 'sms-1', 'status' => 'cancelled'],
            ])),
        ]);
        $resp = $zenvio->sms->cancel('sms-1');
        $this->assertTrue($resp['success']);
        $this->assertSame('cancelled', $resp['data']['status']);
    }

    public function testEmailSend(): void
    {
        $zenvio = $this->getMockClient([
            new Response(200, [], json_encode([
                'success' => true,
                'data' => ['email_ids' => ['email-1'], 'status' => 'queued'],
            ])),
        ]);
        $resp = $zenvio->email->send([
            'from' => 'noreply@example.com',
            'to' => ['u@example.com'],
            'subject' => 'Test',
            'text' => 'Body',
        ]);
        $this->assertSame(['email-1'], $resp['data']['email_ids']);
    }

    public function testEmailCancel(): void
    {
        $zenvio = $this->getMockClient([
            new Response(200, [], json_encode([
                'success' => true,
                'data' => ['email_id' => 'email-1', 'status' => 'cancelled'],
            ])),
        ]);
        $resp = $zenvio->email->cancel('email-1');
        $this->assertTrue($resp['success']);
    }

    public function testMessagesSend(): void
    {
        $zenvio = $this->getMockClient([
            new Response(200, [], json_encode([
                'success' => true,
                'data' => ['message_ids' => ['m1', 'm2'], 'status' => 'queued', 'count' => 2],
            ])),
        ]);
        $resp = $zenvio->messages->send([
            'to' => ['5511999999999'],
            'template' => 'welcome',
            'variables' => ['name' => 'User'],
            'channels' => ['whatsapp', 'sms'],
            'instance_id' => 'inst-1',
        ]);
        $this->assertSame(['m1', 'm2'], $resp['data']['message_ids']);
    }
}

File: packages/sdk-elixir/test/zenvio_test.exs
defmodule ZenvioTest do
  use ExUnit.Case

  setup do
    client = Zenvio.new("test-key")
    {:ok, client: client}
  end

  test "client initialization", %{client: client} do
    assert client.api_key == "test-key"
    assert client.base_url == "https://api.zenvio.com/v1"
  end

  describe "WhatsApp" do
    test "send_text success", %{client: client} do
      adapter = fn %Req.Request{} = req ->
        assert req.options[:method] == :post
        assert to_string(req.url) =~ "whatsapp/send"
        body = req.options[:json]
        assert body["instance_id"] == "instance-1"
        assert body["payload"]["message"] == "Hello"
        Req.Response.new(status: 200, body: %{"message_ids" => ["msg-123"], "status" => "queued"})
      end

      {:ok, response} = Zenvio.Whatsapp.send_text(client, "instance-1", "5511999999999", "Hello", adapter: adapter)
      assert response["message_ids"] == ["msg-123"]
      assert response["status"] == "queued"
    end

    test "send_text error returns status and body", %{client: client} do
      adapter = fn _req ->
        Req.Response.new(status: 400, body: %{"error" => "Invalid instance"})
      end

      assert {:error, %{status: 400, body: body}} = Zenvio.Whatsapp.send_text(client, "x", "1", "hi", adapter: adapter)
      assert body["error"] == "Invalid instance"
    end

    test "send with params", %{client: client} do
      adapter = fn _req -> Req.Response.new(status: 200, body: %{"message_ids" => ["m1"], "status" => "queued"}) end

      params = %{
        "to" => ["5511888888888"],
        "type" => "text",
        "payload" => %{"message" => "Hi"}
      }

      {:ok, resp} = Zenvio.Whatsapp.send(client, "instance-1", params, adapter: adapter)
      assert resp["message_ids"] == ["m1"]
    end

    test "get_message", %{client: client} do
      adapter = fn %Req.Request{} = req ->
        assert to_string(req.url) =~ "whatsapp/msg-1"
        Req.Response.new(status: 200, body: %{"message_id" => "msg-1", "status" => "delivered"})
      end

      {:ok, status} = Zenvio.Whatsapp.get_message(client, "msg-1", adapter: adapter)
      assert status["message_id"] == "msg-1"
      assert status["status"] == "delivered"
    end

    test "list_instances", %{client: client} do
      adapter = fn _req -> Req.Response.new(status: 200, body: %{"success" => true, "data" => [], "pagination" => %{}}) end
      {:ok, list} = Zenvio.Whatsapp.list_instances(client, %{}, adapter: adapter)
      assert list["success"] == true
    end
  end

  describe "SMS" do
    test "send", %{client: client} do
      adapter = fn %Req.Request{} = req ->
        assert to_string(req.url) =~ "sms/send"
        assert req.options[:json]["to"] == ["5511999999999"]
        Req.Response.new(status: 200, body: %{"success" => true, "data" => %{"sms_ids" => ["sms-1"], "status" => "queued"}})
      end

      {:ok, resp} = Zenvio.Sms.send(client, %{"to" => ["5511999999999"], "message" => "Test"}, adapter: adapter)
      assert resp["data"]["sms_ids"] == ["sms-1"]
    end

    test "get", %{client: client} do
      adapter = fn _req -> Req.Response.new(status: 200, body: %{"success" => true, "data" => %{"sms_id" => "sms-1", "status" => "delivered"}}) end
      {:ok, resp} = Zenvio.Sms.get(client, "sms-1", adapter: adapter)
      assert resp["data"]["sms_id"] == "sms-1"
    end
  end

  describe "Email" do
    test "send", %{client: client} do
      adapter = fn _req ->
        Req.Response.new(status: 200, body: %{"success" => true, "data" => %{"email_ids" => ["email-1"], "status" => "queued"}})
      end

      params = %{"from" => "noreply@example.com", "to" => ["u@example.com"], "subject" => "Test", "text" => "Body"}
      {:ok, resp} = Zenvio.Email.send(client, params, adapter: adapter)
      assert resp["data"]["email_ids"] == ["email-1"]
    end

    test "cancel", %{client: client} do
      adapter = fn _req -> Req.Response.new(status: 200, body: %{"success" => true, "data" => %{"email_id" => "email-1", "status" => "cancelled"}}) end
      {:ok, resp} = Zenvio.Email.cancel(client, "email-1", adapter: adapter)
      assert resp["success"] == true
    end
  end

  describe "Messages" do
    test "send", %{client: client} do
      adapter = fn %Req.Request{} = req ->
        assert to_string(req.url) =~ "templates/send"
        body = req.options[:json]
        assert body["template"] == "welcome"
        assert body["channels"] == ["whatsapp", "sms"]
        Req.Response.new(status: 200, body: %{"success" => true, "data" => %{"message_ids" => ["m1", "m2"], "status" => "queued"}})
      end

      params = %{
        "to" => ["5511999999999"],
        "template" => "welcome",
        "variables" => %{"name" => "User"},
        "channels" => ["whatsapp", "sms"],
        "instance_id" => "inst-1"
      }

      {:ok, resp} = Zenvio.Messages.send(client, params, adapter: adapter)
      assert resp["data"]["message_ids"] == ["m1", "m2"]
    end
  end
end
